src/proxy/http/balance/fallback.rs

```diff
@@ -0,0 +1,409 @@
+
+    fn poll_data(&mut self) -> Poll<Option<Self::Data>, Self::Error> {
+        match self {
+            Body::A(ref mut body) => body.poll_data().map(|r| r.map(|o| o.map(Body::A))),
+            Body::B(ref mut body) => body.poll_data().map(|r| r.map(|o| o.map(Body::B))),
+        }
+    }
+
+    fn poll_trailers(&mut self) -> Poll<Option<http::HeaderMap>, Self::Error> {
+        match self {
+            Body::A(ref mut body) => body.poll_trailers(),
+            Body::B(ref mut body) => body.poll_trailers(),
+        }
+    }
+
+    fn is_end_stream(&self) -> bool {
+        match self {
+            Body::A(ref body) => body.is_end_stream(),
+            Body::B(ref body) => body.is_end_stream(),
+        }
+    }
+}
+
+impl<A, B: Default> Default for Body<A, B> {
+    fn default() -> Self {
+        Body::B(Default::default())
+    }
+}
+
+impl<A, B> Body<A, B>
+where
+    A: Payload,
+    B: Payload<Error = A::Error>,
+{
+    fn rsp_a(rsp: http::Response<A>) -> http::Response<Self> {
+        rsp.map(Body::A)
+    }
+
+    fn rsp_b(rsp: http::Response<B>) -> http::Response<Self> {
+        rsp.map(Body::B)
+    }
+}
+
+impl<A, B> Buf for Body<A, B>
+where
+    A: Buf,
+    B: Buf,
+{
+    fn remaining(&self) -> usize {
+        match self {
+            Body::A(ref buf) => buf.remaining(),
+            Body::B(ref buf) => buf.remaining(),
+        }
+    }
+
+    fn bytes(&self) -> &[u8] {
+        match self {
+            Body::A(ref buf) => buf.bytes(),
+            Body::B(ref buf) => buf.bytes(),
+        }
+    }
+
+    fn advance(&mut self, cnt: usize) {
+        match self {
+            Body::A(ref mut buf) => buf.advance(cnt),
+            Body::B(ref mut buf) => buf.advance(cnt),
+        }
+    }
+}
+
+impl<F> Fallback<F>
+where
+    F: rt::Make<router::Config>,
+{
+    fn destroy(&mut self) {
+        self.router = None;
+    }
+
+    fn create(&mut self) {
+        if self.router.is_none() {
+            trace!("creating fallback router...");
+            self.router = Some(self.mk.make(&self.cfg));
+        }
+    }
+}
+
+impl<F, A> svc::Service<http::Request<A>> for Fallback<F>
+where
+    F: rt::Make<router::Config>,
+    F::Value: svc::Service<http::Request<A>>,
+{
+    type Future = <F::Value as svc::Service<http::Request<A>>>::Future;
+    type Error = <F::Value as svc::Service<http::Request<A>>>::Error;
+    type Response = <F::Value as svc::Service<http::Request<A>>>::Response;
+
+    fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+        loop {
+            if let Some(ref mut router) = self.router {
+                return svc::Service::poll_ready(router);
+            } else {
+                self.create();
+            }
+        }
+    }
+
+    fn call(&mut self, req: http::Request<A>) -> Self::Future {
+        loop {
+            if let Some(ref mut router) = self.router {
+                return svc::Service::call(router, req);
+            } else {
+                self.create();
```

Comment:
These seems problematic, as we'll loop into call without having invoked poll_ready. In fact, this seems unreachable.

Reply:
the intention was to guard against a situation where the no endpoints is received between `poll_ready` and `call`. I'm not sure if that's actually a concern. but i think you're right that this is incorrect --- the state transition should only ever occur in poll_ready. 

in the current situation I don't think this is a problem since routers are always ready? but if that changes this would be a problem